checkov/terraform_json/parser.py

```diff
@@ -0,0 +1,199 @@
+from __future__ import annotations
+
+import itertools
+import logging
+from pathlib import Path
+from typing import Any
+
+from charset_normalizer import from_path
+from yaml.scanner import ScannerError
+from yaml import YAMLError
+
+from checkov.common.parsers.json import parse as json_parse
+from checkov.common.parsers.yaml import loader
+from checkov.common.util.consts import LINE_FIELD_NAMES
+from checkov.terraform.graph_builder.graph_components.block_types import BlockType
+
+COMMENT_FIELD_NAME = "//"
+IGNORE_FILED_NAMES = {COMMENT_FIELD_NAME} | LINE_FIELD_NAMES
+SIMPLE_TYPES = (str, int, float, bool)
+
+logger = logging.getLogger(__name__)
+
+
+def parse(file_path: Path) -> tuple[dict[str, Any], list[tuple[int, str]]] | tuple[None, None]:
+    """Parse file to dict object"""
+
+    template = None
+    template_lines = None
+    try:
+        template, template_lines = load(file_path=file_path)
+    except IOError as e:
+        if e.errno == 2:
+            logger.error(f"Template file not found: {file_path}")
+        elif e.errno == 21:
+            logger.error(f"Template references a directory, not a file: {file_path}")
+        elif e.errno == 13:
+            logger.error(f"Permission denied when accessing template file: {file_path}")
+    except UnicodeDecodeError:
+        logger.error(f"Cannot read file contents: {file_path}")
+    except ScannerError as err:
+        if err.problem in ("found character '\\t' that cannot start any token", "found unknown escape character"):
+            try:
+                result = json_parse(file_path, allow_nulls=False)
+                if result:
+                    template, template_lines = result  # type:ignore[assignment]  # this is handled by the next line
+                    if isinstance(template, list):
+                        # should not happen and is more relevant for type safety
+                        template = template[0]
+            except Exception:
+                logger.error(f"Template {file_path} is malformed: {err.problem}")
+                logger.error(f"Tried to parse {file_path} as JSON", exc_info=True)
+    except YAMLError:
+        pass
+
+    if template is None or template_lines is None:
+        return None, None
+
+    return template, template_lines
+
+
+def load(file_path: Path) -> tuple[dict[str, Any], list[tuple[int, str]]]:
```

Comment:
2 questions here:
1. why is this needed instead of just using `json.load`?
2. It's name is confusing as I expected it to be `json.load`, so if it shouldn't be like that I would expect the name to reflect what is special about this load function. Also it is a saved word in `ipython` which added to my confusion as I initially thought this is a general function 😅

Reply:
we need line numbers in the returned `dict`, which normal `json.load(s)` can't offer. Sure, I will rename it, it is somehow named that way in other parsers, so I didn't think much about it 😄